pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java

```diff
@@ -436,6 +449,17 @@ public BrokerResponse handleRequest(JsonNode request, @Nullable RequesterIdentit
     return brokerResponse;
   }
 
+  private void validateNumSegments(Map<ServerInstance, List<String>> offlineRoutingTable,
+      Map<ServerInstance, List<String>> realtimeRoutingTable, int querySegmentLimit) {
+    int numOffline = offlineRoutingTable == null ? 0 : offlineRoutingTable.values().stream().mapToInt(List::size).sum();
```

Comment:
> Avoid using stream api for performance concern

Agreed. Used to be that we had a class instead of a `Map<ServerInstance, List<String>>` for routing table? When the routing table is constructed, we can count the segments and keep it in the class (if we resurrect it), but I go back to my previous question about limiting the segments